File: DynamicProgramming/97_InterleavingString.java
public class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int len1 = s1.length();
        int len2 = s2.length();
        if (len1 + len2 != s3.length()) {
            return false;
        }
        int[] r = new int[len2+1];
        r[0] = 1;
        for (int j=0; j<len2; j++) {
            if (r[j]==1 && s2.charAt(j)==s3.charAt(j)) {
                r[j+1] = 1;
            }
        }

        for (int i=1; i<len1; i++) {
            for (int j=0; j<len2; j++) {
                int first = 1;
                int second = 1;
                if (!(r[j]==1 && s1.charAt(i-1) == s3.charAt(i+j-1))) {
                    first = 0;
                }
                if (!(j>=1 && r[j-1]==1 && s2.charAt(j-1) == s3.charAt(i+j-1))) {
                    second = 0;
                }
                r[j] = first | second;
            }
        }
        return r[len2]==1 ? true : false;
    }
}

File: DynamicProgramming/97_InterleavingString.py
class Solution(object):

    # Time Limit Exceeded
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        len1 = len(s1)
        len2 = len(s2)
        len3 = len(s3)
        if len3 != len1 + len2:
            return False
        # init len2 * len1 rectancle
        r = [[0 for x in range(len1 + 1)] for y in range(len2 + 1)]
        r[0][0] = 1
        # first line
        for j in range(0, len1):
            if r[0][j] == 1 and s1[j] == s3[j]:
                r[0][j + 1] = 1

        # first column
        for i in range(len2):
            if r[i][0] == 1 and s2[i] = s3[i]:
                r[i + 1][0] = 1

        # the contents of matrix
        for i in range(1, len2 + 1):
            for j in range(1, len1 + 1):
                if r[i - 1][j] == 1:
                    if s2[i - 1] == s3[i + j - 1]:
                        r[i][j] = 1
                if r[i][j - 1] == 1:
                    if s1[j - 1] == s3[i + j - 1]:
                        r[i][j] = 1

        return r[len2][len1] == 1
